index.js

```diff
@@ -4,79 +4,153 @@ var React = require('react');
-      //
-      // We check for `== null` to cover both `null` and `undefined`, this is
-      // what React does to check if component is used in uncontrolled mode.
-      this.props.value == null
-    ) {
-      this.recalculateSize();
-    }
-  },
+    displayName: 'TextareaAutosize',
 
-  onChange: function(e) {
-    if (this.props.onChange) {
-      this.props.onChange(e);
-    }
-    if (this.props.value === undefined) {
-      // controlled mode
-      this.recalculateSize();
-    }
-  },
-
-  recalculateSize: function() {
-    var diff;
-    var node = this.getDOMNode();
-
-    if (window.getComputedStyle) {
-      var styles = window.getComputedStyle(node);
-
-      // If the textarea is set to border-box, it's not necessary to
-      // subtract the padding.
-      if (styles.getPropertyValue('box-sizing') === "border-box" ||
-          styles.getPropertyValue('-moz-box-sizing') === "border-box" ||
-          styles.getPropertyValue('-webkit-box-sizing') === "border-box") {
-        diff = 0;
-      } else {
-        diff = (
-          parseInt(styles.getPropertyValue('padding-bottom') || 0, 10) +
-          parseInt(styles.getPropertyValue('padding-top') || 0, 10)
-        );
-      }
-    } else {
-      diff = 0;
-    }
+    propTypes: {
+        minHeight: React.PropTypes.number,
+        maxHeight: React.PropTypes.number
+    },
+
+    /**
+     * This component can work with the ValueLink shortcut provided by React.
+     * When ValueLink is passed, it tries to take over the onChange event in the underlying element
+     * However, we need onChange to calculate the height of the box ourselves
+     * So we store ValueLink locally, remove it from the underlying <textarea>, but proxy all onChange calls to it
+     */
+    valueLink: undefined,
+
+    /**
+     * A cached copy of properties that will be passed to the underlying <textarea>
+     * Essentially passes all own properties, but sanitizes them to remove ValueLink and add the necessary style settings
+     */
+    textareaProps: {},
+
+    getInitialState: function () {
+        return {
+            height:   'auto',
+            overflow: 'hidden'
+        }
+    },
+
+    render: function () {
+        this.textareaProps.style = objectAssign({}, this.textareaProps.style, {
+            height:   this.state.height,
+            overflow: this.state.overflow
+        });
+
+        return React.DOM.textarea(this.textareaProps);
+    },
+
+    componentWillMount: function () {
+        this.textareaProps = this.sanitizeProps(this.props);
+    },
+
+
+    /**
+     * Do the initial calcluation of the height
+     * And set an event listener to update the textarea height when the window is resized
+     */
+    componentDidMount: function () {
+        //do a short timeout so let the browser sort out its paddings etc
+        //otherwise sometimes the browser doesn't not text should be wrapped and missed some cases
+        setTimeout(this.recalculateSize);
```

Comment:
1ms should be quick enough for the jump not to be noticeable, but sometimes browsers misbehave with timeout and yes you might see a jump.

This particular implementation is quite naive because it directly manipulates DOM, and that poses two problems: page scrolling is affected and it is not very React-idiomatic. Additionally there is a performance hit from calculating computedStyle on each key stroke.

I didn't end up using this code and wrote another one specifically fitted to our needs, but didn't have a chance to prepare multiple PR for this project yet. I'll post up our version shortly for you to review & compare.